File: creme/sms/models/template.py
################################################################################
#    Creme is a free/open-source Customer Relationship Management software
#    Copyright (C) 2009-2022  Hybird
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU Affero General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Affero General Public License for more details.
#
#    You should have received a copy of the GNU Affero General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
################################################################################

from django.core.exceptions import ValidationError
from django.db import models
from django.urls import reverse
from django.utils.functional import lazy
from django.utils.translation import gettext
from django.utils.translation import gettext_lazy as _
from django.utils.translation import pgettext_lazy

from creme.creme_core.models import CremeEntity

from ..encoding import SMS_MAX_LENGTH, gsm_encoded_content

SPECIAL_CHARS = '^ { } \\ [ ~ ] | €'  # TODO: given by the backend ?


class AbstractMessageTemplate(CremeEntity):
    name = models.CharField(_('Name'), max_length=100)
    subject = models.CharField(_('Subject'), max_length=100)
    body = models.TextField(
        # _('Body'),
        pgettext_lazy('sms', 'Message body'),
        help_text=lazy(
            (lambda: gettext(
                'Message with a maximum of 160 characters.\n'
                'Beware, the header matters (+ 3 characters) '
                'and the following characters count double: {}'
            ).format(SPECIAL_CHARS)),
            str
        )()
    )

    error_messages = {
        'too_long': _('Message is too long (%(length)s > %(max_length)s)'),
    }

    creation_label = pgettext_lazy('sms-template', 'Create a template')
    save_label     = pgettext_lazy('sms-template', 'Save the template')

    class Meta:
        abstract = True
        app_label = 'sms'
        verbose_name = _('SMS Message template')
        verbose_name_plural = _('SMS Messages templates')
        ordering = ('name',)

    def __str__(self):
        return self.name

    def clean(self):
        subject = self.subject
        # TODO: factorise
        content = f'{subject} : {self.body}' if subject else self.body
        encoded_length = len(gsm_encoded_content(content))

        if encoded_length > SMS_MAX_LENGTH:
            raise ValidationError(
                self.error_messages['too_long'],
                params={
                    'length':     encoded_length,
                    'max_length': SMS_MAX_LENGTH,
                },
                code='too_long',
            )

    def get_absolute_url(self):
        return reverse('sms__view_template', args=(self.id,))

    @staticmethod
    def get_create_absolute_url():
        return reverse('sms__create_template')

    def get_edit_absolute_url(self):
        return reverse('sms__edit_template', args=(self.id,))

    @staticmethod
    def get_lv_absolute_url():
        return reverse('sms__list_templates')

    # def resolve(self, date):
    #     return self.subject + ' : ' + self.body


class MessageTemplate(AbstractMessageTemplate):
    class Meta(AbstractMessageTemplate.Meta):
        swappable = 'SMS_TEMPLATE_MODEL'

File: creme/creme_core/static/creme_core/js/widgets/utils/template.js
/*******************************************************************************
    Creme is a free/open-source Customer Relationship Management software
    Copyright (C) 2009-2015  Hybird

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*******************************************************************************/

(function($) {
"use strict";

creme.utils = creme.utils || {};

creme.utils.TemplateRenderer = creme.component.Component.sub({
    tags: function(template) { return []; },
    render: function(template, values) { return template; }
});

creme.utils.TemplateDefaultRenderer = creme.utils.TemplateRenderer.sub({
    _entrypattern: /\$\{([\w\d\\.^\\s]+)\}/g,

    _attributeGetter: function(data, key) {
        var keys = Array.isArray(key) ? key : ((typeof key === 'string') ? key.split('.') : [key]);
        var value = data;

        for (var i in keys) {
            if (!Object.isNone(value) && typeof value === 'object') {
                value = value[keys[i]];
            } else {
                value = undefined;
                break;
            }

            value = Object.isFunc(value) ? value(key) : value;
        }

        return value;
    },

    tags: function(template) {
        var matches = template.match(this._entrypattern);
        var tags = {};

        if (Object.isEmpty(matches) === false) {
            matches.forEach(function(match) {
                tags[match.slice(2, -1).split('.')[0]] = 0;
                return 0;
            });
        }

        return Object.keys(tags);
    },

    render: function(template, values) {
        if (Object.isEmpty(values)) {
            return template;
        }

        var getter = this._attributeGetter.bind(this);

        return template.replace(this._entrypattern, function(match, key) {
            var value = getter(values, key);
            return value !== undefined ? value : match;
        });
    }
});

creme.utils.Template = creme.component.Component.sub({
    _init_: function(pattern, parameters, renderer) {
        this.renderer(renderer || new creme.utils.TemplateDefaultRenderer());
        this.pattern(pattern);
        this.parameters(parameters);
    },

    _resolve: function(extra) {
        extra = extra || {};

        var data = this._parameters || {};
        var resolved = {};

        if (Object.isFunc(data)) {
            this.tags().forEach(function(key) {
                resolved[key] = data(key);
            });
        } else {
            resolved = $.extend(resolved, data);
        }

        if (Object.isFunc(extra)) {
            this.tags().forEach(function(key) {
                resolved[key] = extra(key);
            });
        } else {
            resolved = $.extend(resolved, extra);
        }

        return resolved;
    },

    render: function(extra) {
        var renderer = this._renderer;
        var pattern = this._pattern;

        if (Object.isNone(renderer) || Object.isNone(pattern)) {
            return null;
        }

        return renderer.render(pattern, this._resolve(extra));
    },

    tags: function() {
        return this._tags;
    },

    _updateTags: function() {
        var renderer = this._renderer;
        var pattern = this._pattern;

        this._tags = (Object.isNone(renderer) || Object.isEmpty(pattern)) ? [] : renderer.tags(pattern);
    },

    iscomplete: function() {
        var tags = this._tags;
        var parameters = this._resolve();

        for (var i = 0; i < tags.length; ++i) {
            if (Object.isNone(parameters[tags[i]])) {
                return false;
            }
        }

        return true;
    },

    renderer: function(renderer) {
        if (renderer === undefined) {
            return this._renderer;
        }

        this._renderer = renderer;
        this._updateTags();
        return this;
    },

    pattern: function(pattern) {
        if (pattern === undefined) {
            return this._pattern;
        }

        this._pattern = pattern;
        this._updateTags();
        return this;
    },

    parameters: function(parameters) {
        return Object.property(this, '_parameters', parameters);
    },

    update: function(data) {
        // data is a string, use it as url
        if (Array.isArray(data)) {
            this.pattern(data[0]);
            this.parameters(data[1]);
        } else if (typeof data === 'object') {
            this.parameters($.extend({}, this.parameters() || {}, data));
        } else if (typeof data === 'string') {
            this.pattern(data);
        }

        return this;
    }
});

creme.utils.templatize = function(value, context) {
    var template;

    if (Object.isNone(value)) {
        template = new creme.utils.Template();
    }

    if (Object.isType(value, 'string')) {
        template = new creme.utils.Template(value);
    }

    if (value !== null && Object.isType(value, 'object') && Object.isFunc(value.is) && value.is(creme.utils.Template)) {
        template = value;
    }

    return Object.isNone(context) ? template : template.parameters(context);
};
}(jQuery));
